service/history/replication/dlq_handler_test.go

```diff
@@ -110,11 +113,176 @@ func (s *dlqHandlerSuite) TearDownTest() {
+	tests := []struct {
+		name   string
+		status int32
+	}{
+		{
+			name:   "stopped",
+			status: common.DaemonStatusStopped,
+		},
+		{
+			name:   "started",
+			status: common.DaemonStatusStarted,
+		},
+	}
+
+	for _, tc := range tests {
+		s.T().Run(tc.name, func(t *testing.T) {
+			s.messageHandler.status = tc.status
+
+			s.messageHandler.Stop()
+		})
+	}
+}
+
+func (s *dlqHandlerSuite) TestGetMessageCount() {
+	size := int64(1)
+	tests := []struct {
+		name         string
+		latestCounts map[string]int64
+		forceFetch   bool
+		err          error
+	}{
+		{
+			name:         "success",
+			latestCounts: map[string]int64{s.sourceCluster: size},
+		},
+		{
+			name:       "success with fetchAndEmitMessageCount call",
+			forceFetch: true,
+		},
+		{
+			name:       "error",
+			forceFetch: true,
+			err:        errors.New("fetchAndEmitMessageCount error"),
+		},
+	}
+
+	for _, tc := range tests {
+		s.T().Run(tc.name, func(t *testing.T) {
+			s.messageHandler.latestCounts = tc.latestCounts
+
+			if tc.forceFetch || tc.latestCounts == nil {
+				s.executionManager.On("GetReplicationDLQSize", mock.Anything, mock.Anything).Return(&persistence.GetReplicationDLQSizeResponse{Size: size}, tc.err).Times(1)
+			}
+
+			counts, err := s.messageHandler.GetMessageCount(context.Background(), tc.forceFetch)
+
+			if tc.err != nil {
+				s.Error(err)
+				s.Equal(tc.err, err)
+			} else if tc.latestCounts != nil {
+				s.NoError(err)
+				s.Equal(size, counts[s.sourceCluster])
+			} else {
+				s.NoError(err)
+			}
+		})
+	}
+}
+
+func (s *dlqHandlerSuite) TestFetchAndEmitMessageCount() {
+	tests := []struct {
+		name string
+		err  error
+	}{
+		{
+			name: "success",
+			err:  nil,
+		},
+		{
+			name: "error",
+			err:  errors.New("error"),
+		},
+	}
+
+	for _, tc := range tests {
+		s.T().Run(tc.name, func(t *testing.T) {
+			size := int64(3)
+			rets := &persistence.GetReplicationDLQSizeResponse{Size: size}
+			s.messageHandler.latestCounts = make(map[string]int64)
+
+			s.executionManager.On("GetReplicationDLQSize", context.Background(), mock.Anything).Return(rets, tc.err).Times(1)
+
+			err := s.messageHandler.fetchAndEmitMessageCount(context.Background())
+
+			if tc.err != nil {
+				s.Error(err)
+				s.Equal(tc.err, err)
+			} else {
+				s.NoError(err)
+				s.Equal(len(s.messageHandler.latestCounts), len(s.taskExecutors))
+				s.Equal(size, s.messageHandler.latestCounts[s.sourceCluster])
+			}
+		})
+	}
+}
+
+func (s *dlqHandlerSuite) TestEmitDLQSizeMetricsLoop_StopsWhenDoneSignalReceived() {
+	s.messageHandler.status = common.DaemonStatusStarted
+	go s.messageHandler.emitDLQSizeMetricsLoop()
+
+	s.messageHandler.Stop()
+	time.Sleep(1000 * time.Millisecond) // Allow go routine to process done signal
+
+	s.Equal(common.DaemonStatusStopped, s.messageHandler.status)
+}
+
+func (s *dlqHandlerSuite) TestEmitDLQSizeMetricsLoop_FetchesAndEmitsMetricsPeriodically() {
+	emissionNumber := 2
+	interval := 1 * time.Second
```

Comment:
you can lower this to 100ms or something to lower the overall test run time

Reply:
Made an attempt to use timeSource and Advance. Let me know what you think. Had to make some changes to the methods to avoid using `sleep`, as suggested by @3vilhamster